amethyst_input/src/input_handler.rs

```diff
@@ -298,4 +474,28 @@ where
             .get(action)
             .map(|ref buttons| buttons.iter().any(|&b| self.button_is_down(b)))
     }
+
+    /// Retrieve next free controller number to allocate new controller to
+    fn alloc_controller_id(&self) -> u32 {
+        let mut i = 0u32;
+        loop {
+            if self
+                .connected_controllers
+                .iter()
+                .find(|ids| ids.0 == i)
+                .is_none()
+            {
+                return i;
+            }
+            i += 1;
```

Comment:
If list order is preserved, you can just get the last controller in the list and return id+1

Reply:
Plug 1
Plug 2
Bind 1 and 2
Unplug 1
Unplug 2
Plug 2
Plug 1
Broken.

I prefer to use incremental to show that this isn't the same connection, and to indicate that the old one has been dropped. (but my method doesn't do that with a single connection anyway).

I guess that's fine like it is anyway.

Reply:
You will still have a working game that way. Simply controllers will be switched for every player. Well, player will have to switch them in real world i guess 😉